### ISSUE TITLE
False positive with auto-dereferencing

### ISSUE THREAD

**Issue by StragaSevera** [intellij-rust/intellij-rust#8856]
## Environment

* **IntelliJ Rust plugin version:** 0.4.170.4627-221
* **Rust toolchain version:** 1.61.0 (fe5b13d68 2022-05-18) x86_64-unknown-linux-gnu
* **IDE name and version:** CLion 2022.1.1 (CL-221.5591.52)
* **Operating system:** Linux 5.17.7-arch1-1
* **Macro expansion engine:** new
* **Name resolution engine:** new

## Problem description
The code I'm working on uses auto-dereferencing for calling a method on a immutable reference. Neither the compiler nor Clippy find any errors, but IntelliJ reports one:

![image](https://user-images.githubusercontent.com/6378091/169353899-f45f0fdb-56a5-43b2-83f4-07f58c1cbdfd.png)

## Steps to reproduce

I failed to make a small example that can show the error, sorry. Here is the link to the [repository](https://github.com/StragaSevera/minesweeper-tutorial/tree/a934dbfc5411b40bd72943a741d2bf525ad50d66) with the false positive, the file is `board_plugin/src/lib.rs`.
Snippet from the code:
```rust
    fn insert_bomb(size: f32, padding: f32, bomb_image: &Handle<Image>, cmd: &mut EntityCommands) {
        cmd.insert(Bomb);
        cmd.with_children(|parent| {
            parent.spawn_bundle(SpriteBundle {
                sprite: Sprite {
                    custom_size: Some(Vec2::splat(size - padding)),
                    ..Default::default()
                },
                transform: Transform::from_xyz(0., 0., 1.),
                texture: bomb_image.clone(),
                ..Default::default()
            });
        });
    }
```

**Comment by Undin** [comment]
I can reproduce it with the latest stable plugin but not with nightly one.
So it seems it's already fixed by recent changes in type inference engine.
Could you check it yourself? See this [instruction](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-quick-start.html#install-nightly) for installing nightly plugin build

**Comment by StragaSevera** [comment]
No, it still shows a false positive, even on a nightly build.

**Comment by Undin** [comment]
@StragaSevera oh, sorry, I missed that the pointed project uses procedural macros (because of `bevy` at least) and my IDE setup with stable plugin has default settings. In particular, the plugin knows nothing about result of proc macro expansion by default.
And here proper selection of `clone()` implementation depends on procedural macro expansion. So with the default setup, the plugin incorrectly finds `clone` method implementation, incorrectly infer the type, and you get false positive type mismatch error

But you can turn on our experimental procedural macro expansion support. See #6908 for more details

It actually doesn't depend on stable or nightly plugin versions, so you can continue using stable one if you want

**Comment by IlyaFedorinov2** [comment]
It works with experimental macro expansions, thanks! =-)

**Comment by StragaSevera** [comment]
It works with experimental macro expansions, thanks! =-)
